internal/resources/webform/webform.js

```diff
@@ -2354,6 +2404,115 @@ window.initGRPCForm = function(services, invokeURI, metadataURI, debug) {
         }
     }
 
+    let history = [];
+    const maxHistory = 100;
+    const target = $(".target").text();
+
+    const storageKey = `grpcui-history-${window.location.host}-${target}`;
+
+    const loadHistory = () => {
+        const json = localStorage.getItem(storageKey);
+        if (json) {
+            history = (JSON.parse(json))
+        }
+        updateHistoryUI();
+    }
+
+    const writeHistory = () => {
+        localStorage.setItem(storageKey, JSON.stringify(history));
+    }
+
+    const clearHistory = () => {
+        if(confirm('Are you sure you wish to delete all history? This action is permanent and cannot be undone.')) {
+            history = [];
+            writeHistory();
```

Comment:
All calls to `writeHistory` are always immediately followed by `updateHistoryUI`. Maybe they should be combined -- like rename `writeHistory` to `onHistoryChange` and have it save to local storage and then call `updateHistoryUI`?